File: app/Sources/GunkApp/Extract/Extractor.swift
import Foundation

enum ExtractorError: Error, Equatable {
  case sourceNotFound(Int64)
  case invalidRelativePath(String)
  case sourceFileOutsideRoot(String)
}

struct ExtractionResult: Equatable, Sendable {
  let bundleURL: URL
  let manifestURL: URL
  let readmeURL: URL
  let redactions: [Redaction]
}

final class Extractor {
  static let defaultConfidenceThreshold = 0.7

  private let store: Store
  private let gunkHome: URL
  private let confidenceThreshold: Double
  private let fileManager: FileManager
  private let redactor: SecretRedactor
  private let manifestWriter: ManifestWriter
  private let licenseDetector: LicenseDetector
  private let now: () -> Int64

  init(
    store: Store,
    gunkHome: URL = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".gunk"),
    confidenceThreshold: Double = Extractor.defaultConfidenceThreshold,
    fileManager: FileManager = .default,
    redactor: SecretRedactor = SecretRedactor(),
    manifestWriter: ManifestWriter = ManifestWriter(),
    licenseDetector: LicenseDetector = LicenseDetector(),
    now: @escaping () -> Int64 = Store.currentTimeInMilliseconds
  ) {
    self.store = store
    self.gunkHome = gunkHome.standardizedFileURL
    self.confidenceThreshold = confidenceThreshold
    self.fileManager = fileManager
    self.redactor = redactor
    self.manifestWriter = manifestWriter
    self.licenseDetector = licenseDetector
    self.now = now
  }

  @discardableResult
  func extract(gunk: Gunk) throws -> ExtractionResult? {
    guard (gunk.confidence ?? 0) >= confidenceThreshold else {
      return nil
    }

    guard let source = try store.source(id: gunk.sourceId) else {
      throw ExtractorError.sourceNotFound(gunk.sourceId)
    }

    let sourceRoot = URL(fileURLWithPath: source.path).standardizedFileURL
    let bundleURL = gunkHome
      .appendingPathComponent("modules", isDirectory: true)
      .appendingPathComponent(String(gunk.id), isDirectory: true)
    let manifestURL = bundleURL.appendingPathComponent("gunk.yml")
    let readmeURL = bundleURL.appendingPathComponent("README.gunk.md")
    let gunkFiles = try store.filesForGunk(gunkId: gunk.id)
    let tags = try store.listGunkTags(gunkId: gunk.id).map(\.tag)
    var redactions: [Redaction] = []

    if fileManager.fileExists(atPath: bundleURL.path) {
      try fileManager.removeItem(at: bundleURL)
    }
    try fileManager.createDirectory(at: bundleURL, withIntermediateDirectories: true)

    for file in gunkFiles {
      try copyFile(
        relpath: file.relpath,
        sourceRoot: sourceRoot,
        bundleRoot: bundleURL,
        redactions: &redactions
      )
    }

    let extractedAt = now()
    let artifact = manifestWriter.artifact(
      input: ManifestInput(
        gunk: gunk,
        tags: tags,
        files: gunkFiles.map(\.relpath),
        sourcePath: manifestWriter.homeRelativePath(source.path),
        sourceCommit: shortCommitHash(sourceRoot: sourceRoot),
        license: try licenseDetector.detect(sourceRoot: sourceRoot),
        redactions: redactions,
        extractedAt: Date(timeIntervalSince1970: Double(extractedAt) / 1_000)
      )
    )

    try artifact.manifest.write(to: manifestURL, atomically: true, encoding: .utf8)
    try artifact.readme.write(to: readmeURL, atomically: true, encoding: .utf8)
    try store.updateGunkExtraction(
      id: gunk.id,
      bundlePath: bundleURL.path,
      manifestPath: manifestURL.path,
      extractedAt: extractedAt
    )

    return ExtractionResult(
      bundleURL: bundleURL,
      manifestURL: manifestURL,
      readmeURL: readmeURL,
      redactions: redactions
    )
  }

  private func copyFile(
    relpath: String,
    sourceRoot: URL,
    bundleRoot: URL,
    redactions: inout [Redaction]
  ) throws {
    try validateRelativePath(relpath)

    let sourceURL = sourceRoot.appendingPathComponent(relpath).standardizedFileURL
    let sourceRootPath = sourceRoot.path.hasSuffix("/") ? sourceRoot.path : sourceRoot.path + "/"
    guard sourceURL.path.hasPrefix(sourceRootPath) else {
      throw ExtractorError.sourceFileOutsideRoot(relpath)
    }

    let result = try redactor.redact(url: sourceURL, relpath: relpath)
    redactions.append(contentsOf: result.redactions)

    guard case .write(let data, _) = result else {
      return
    }

    let destinationURL = bundleRoot.appendingPathComponent(relpath)
    try fileManager.createDirectory(
      at: destinationURL.deletingLastPathComponent(),
      withIntermediateDirectories: true
    )
    try data.write(to: destinationURL, options: .atomic)
  }

  private func validateRelativePath(_ relpath: String) throws {
    let normalized = relpath.replacingOccurrences(of: "\\", with: "/")

    guard !normalized.hasPrefix("/"),
          !normalized.isEmpty,
          !normalized.split(separator: "/").contains("..") else {
      throw ExtractorError.invalidRelativePath(relpath)
    }
  }

  private func shortCommitHash(sourceRoot: URL) -> String? {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
    process.arguments = ["git", "-C", sourceRoot.path, "rev-parse", "--short", "HEAD"]

    let output = Pipe()
    process.standardOutput = output
    process.standardError = Pipe()

    do {
      try process.run()
      process.waitUntilExit()
    } catch {
      return nil
    }

    guard process.terminationStatus == 0 else {
      return nil
    }

    let data = output.fileHandleForReading.readDataToEndOfFile()
    return String(data: data, encoding: .utf8)?
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .nilIfEmpty
  }
}

private extension String {
  var nilIfEmpty: String? {
    isEmpty ? nil : self
  }
}

File: engine/src/extract/extractor.ts
// Module bundle extractor. Ported from
// app/Sources/GunkApp/Extract/Extractor.swift.

import { spawnSync } from "node:child_process";
import { existsSync, mkdirSync, rmSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join, resolve } from "node:path";

import {
  filesForGunk,
  listGunkTags,
  markGunkExtracted,
  sourceById,
  type Gunk,
} from "../store/index.js";
import type { Database } from "bun:sqlite";
import { LicenseDetector } from "./licenseDetector.js";
import { ManifestWriter } from "./manifestWriter.js";
import { SecretRedactor, type Redaction } from "./secretRedactor.js";

export interface ExtractionResult {
  bundlePath: string;
  manifestPath: string;
  readmePath: string;
  redactions: Redaction[];
}

export class ExtractorError extends Error {
  constructor(
    message: string,
    readonly code: "sourceNotFound" | "invalidRelativePath" | "sourceFileOutsideRoot",
  ) {
    super(message);
    this.name = "ExtractorError";
  }
}

export interface ExtractorOptions {
  gunkHome?: string;
  confidenceThreshold?: number;
  redactor?: SecretRedactor;
  manifestWriter?: ManifestWriter;
  licenseDetector?: LicenseDetector;
  now?: () => number;
}

export class Extractor {
  static readonly defaultConfidenceThreshold = 0.7;

  private readonly gunkHome: string;
  private readonly confidenceThreshold: number;
  private readonly redactor: SecretRedactor;
  private readonly manifestWriter: ManifestWriter;
  private readonly licenseDetector: LicenseDetector;
  private readonly now: () => number;

  constructor(
    private readonly db: Database,
    options: ExtractorOptions = {},
  ) {
    this.gunkHome = resolve(options.gunkHome ?? join(homedir(), ".gunk"));
    this.confidenceThreshold = options.confidenceThreshold ?? Extractor.defaultConfidenceThreshold;
    this.redactor = options.redactor ?? new SecretRedactor();
    this.manifestWriter = options.manifestWriter ?? new ManifestWriter();
    this.licenseDetector = options.licenseDetector ?? new LicenseDetector();
    this.now = options.now ?? Date.now;
  }

  extract(gunk: Gunk): ExtractionResult | null {
    if ((gunk.confidence ?? 0) < this.confidenceThreshold) {
      return null;
    }
    const source = sourceById(this.db, gunk.sourceId);
    if (!source) {
      throw new ExtractorError(`Source ${gunk.sourceId} not found.`, "sourceNotFound");
    }

    const sourceRoot = resolve(source.path);
    const bundlePath = join(this.gunkHome, "modules", String(gunk.id));
    const manifestPath = join(bundlePath, "gunk.yml");
    const readmePath = join(bundlePath, "README.gunk.md");
    const gunkFiles = filesForGunk(this.db, gunk.id);
    const tags = listGunkTags(this.db, gunk.id).map((t) => t.tag);
    const redactions: Redaction[] = [];

    if (existsSync(bundlePath)) {
      rmSync(bundlePath, { recursive: true, force: true });
    }
    mkdirSync(bundlePath, { recursive: true });

    for (const file of gunkFiles) {
      this.copyFile(file.relpath, sourceRoot, bundlePath, redactions);
    }

    const extractedAt = this.now();
    const artifact = this.manifestWriter.artifact({
      gunk,
      tags,
      files: gunkFiles.map((f) => f.relpath),
      sourcePath: this.manifestWriter.homeRelativePath(source.path),
      sourceCommit: this.shortCommitHash(sourceRoot),
      license: this.licenseDetector.detect(sourceRoot),
      redactions,
      extractedAtMs: extractedAt,
    });

    writeFileSync(manifestPath, artifact.manifest, "utf-8");
    writeFileSync(readmePath, artifact.readme, "utf-8");
    markGunkExtracted(this.db, gunk.id, bundlePath, manifestPath, extractedAt);

    return { bundlePath, manifestPath, readmePath, redactions };
  }

  private copyFile(relpath: string, sourceRoot: string, bundleRoot: string, redactions: Redaction[]): void {
    this.validateRelativePath(relpath);
    const sourceURL = resolve(sourceRoot, relpath);
    const sourceRootPath = sourceRoot.endsWith("/") ? sourceRoot : `${sourceRoot}/`;
    if (!sourceURL.startsWith(sourceRootPath)) {
      throw new ExtractorError(`Source file outside root: ${relpath}`, "sourceFileOutsideRoot");
    }
    const result = this.redactor.redact(sourceURL, relpath);
    redactions.push(...result.redactions);
    if (result.kind !== "write") return;
    const destination = join(bundleRoot, relpath);
    mkdirSync(dirname(destination), { recursive: true });
    writeFileSync(destination, result.data);
  }

  private validateRelativePath(relpath: string): void {
    const normalized = relpath.replace(/\\/g, "/");
    if (normalized.startsWith("/") || normalized.length === 0 || normalized.split("/").includes("..")) {
      throw new ExtractorError(`Invalid relative path: ${relpath}`, "invalidRelativePath");
    }
  }

  private shortCommitHash(sourceRoot: string): string | null {
    try {
      const result = spawnSync("git", ["-C", sourceRoot, "rev-parse", "--short", "HEAD"], { encoding: "utf-8" });
      if (result.status !== 0) return null;
      const hash = result.stdout.trim();
      return hash.length === 0 ? null : hash;
    } catch {
      return null;
    }
  }
}
